File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.scala
object Solution {
    def findSubstring(s: String, words: Array[String]): List[Int] = {
        if (s.isEmpty || words.isEmpty) return List()
        val wordLen = words(0).length
        val wordCount = words.groupBy(identity).view.mapValues(_.length).toMap
        val totalLen = wordLen * words.length
        val res = scala.collection.mutable.ListBuffer[Int]()
        for (i <- 0 until wordLen) {
            var left = i
            var right = i
            var seen = scala.collection.mutable.Map[String, Int]()
            while (right + wordLen <= s.length) {
                val word = s.substring(right, right + wordLen)
                right += wordLen
                if (wordCount.contains(word)) {
                    seen(word) = seen.getOrElse(word, 0) + 1
                    while (seen(word) > wordCount(word)) {
                        val leftWord = s.substring(left, left + wordLen)
                        seen(leftWord) -= 1
                        left += wordLen
                    }
                    if (right - left == totalLen) {
                        res += left
                    }
                } else {
                    seen.clear()
                    left = right
                }
            }
        }
        res.toList
    }
}

File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.swift
class Solution {
    func findSubstring(_ s: String, _ words: [String]) -> [Int] {
        if words.isEmpty { return [] }
        let wordLen = words[0].count, totalLen = wordLen * words.count
        var wordCount = [String: Int]()
        for w in words { wordCount[w, default: 0] += 1 }
        let sArr = Array(s)
        var res = [Int]()
        for i in 0..<wordLen {
            var left = i, count = 0
            var seen = [String: Int]()
            var j = i
            while j + wordLen <= s.count {
                let word = String(sArr[j..<j+wordLen])
                if let _ = wordCount[word] {
                    seen[word, default: 0] += 1
                    count += 1
                    while seen[word]! > wordCount[word]! {
                        let leftWord = String(sArr[left..<left+wordLen])
                        seen[leftWord]! -= 1
                        left += wordLen
                        count -= 1
                    }
                    if count == words.count { res.append(left) }
                } else {
                    seen.removeAll()
                    count = 0
                    left = j + wordLen
                }
                j += wordLen
            }
        }
        return res
    }
}